include/envoy/network/listen_socket.h

```diff
@@ -90,6 +90,39 @@ class ConnectionSocket {
    * Close the underlying socket.
    */
   virtual void close() PURE;
+
+  /**
+   * Visitor class for setting socket options.
+   */
+  class Options {
+  public:
+    virtual ~Options() {}
+
+    /**
+     * @param socket the socket on which to apply options.
+     * @return true if succeeded, false otherwise.
+     */
+    virtual bool setOptions(ConnectionSocket& socket) const PURE;
+
+    /**
+     * @return bits that can be used to separate upstream connections based on the options. Should
+     *         return zero if connections with different options can be pooled together.
+     */
+    virtual uint32_t hashKey() const PURE;
```

Comment:
At the interface level can we make this uint64_t? Seems like it should just be 64-bit. If the implementation chooses to just use the low 32-bits that's an implementation detail?

Reply:
I was going back and forth with this a bit. The problem with dropping the upper bits is that then wrong connections could be reused. For example, let's say based on the server connections we want to apply options identified with '0x1234' to upstream connections. For another server connection we want to apply options identified with '0xff34' to the upstream connections. If the implementation drops the upper half and only considers '0x34' then it would reuse the first connection in the second case, and the traffic would be forwarded with e.g., wrong socket mark.

Maybe I should not have used the word "hash" in this context. It came from the fact that we use the return value as part of a connection pool hash key.

The cleanest interface I can think of would take in a std::vector<uint8_t> and the visitor can then append as many bytes as it needs to the hash key. Internally we would init it with one byte covering the Http::Protocol and priority. Better?

Reply:
It seems like the user controls both sides, right? So if we do 64-bit, wouldn't the top 32-bits always be zero? I'm confused what the problem is.

Reply:
The other side is in cluster_manager_impl, where we pick a connection pool from the connection pool container. Similar to the HTTP/1 vs. HTTP/2 or priority, we have to keep upstream connections with different options separate, i.e., keep them in separate pools. This piece is not pluggable, so it is not controlled by the filter setting the options.

To achieve the connection separation, I've changed the connection pool container from an array to an unordered_map. For this to work the options (or an encoding of them, depending on what the options are) need to be in the map key.

Right now in the PR I've assembled a uint64_t key from 1 bit for the priority, 2 bits  for the Http::Protocol and 32 bits from here. This works for my intended use case (socket mark), and I see that more bits would be needed if there were some QoS options, ToS bits etc. in the play. But even then, uint64_t for this purpose might be too restrictive.

One version I had locally had this as uint64_t, but with a comment that the top three bits may be ignored by the caller. But imagine if we add more protocols, etc, and have to change it to "top 4 bits being ignored", existing filter implementations could be broken.

See the problem now?

Reply:
Yup fair enough. 32-bit seems fine. I would add more comments explaining why this is the case.